# Shopify Winter 2022 Data Science Intern Challenge

by Vadim Fridman 
vadim.fridman@fu-berlin.de

_**On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis.**_

### Question 1

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("DS1.csv")

In [3]:
#df.info()

In [4]:
df.sample(10)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
2697      2698       73      704           495            3           cash   
832        833       55      780           171            1          debit   
1567      1568       12      753           402            2           cash   
169        170        5      970           284            2          debit   
1702      1703        7      774           336            3          debit   
4829      4830       18      770           156            1    credit_card   
2725      2726       40      783           644            4          debit   
3633      3634       90      999           534            3          debit   
73          74       14      968           116            1           cash   
2388      2389       73      846           330            2    credit_card   

               created_at  
2697  2017-03-17 15:00:43  
832   2017-03-05 12:54:04  
1567  2017-03-28 14:51:36  
169    2017-03-24 0:33:57  
1702   2017-03-13 0:48:28  
4829   2017-03-18 2:19:12  
2725   2017-03-23 9:58:41  
3633  2017-03-01 12:04:51  
73     2017-03-22 0:14:48  
2388   2017-03-14 4:19:59

In [5]:
df.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

In [6]:
df[df["order_amount"] == 704000]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
520        521       42      607        704000         2000    credit_card   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1436      1437       42      607        704000         2000    credit_card   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
2153      2154       42      607        704000         2000    credit_card   
2297      2298       42      607        704000         2000    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
3332      3333       42      607        704000         2000    credit_card   
4056      4057       42      607        704000         2000    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   

              created_at  
15    2017-03-07 4:00:00  
60    2017-03-04 4:00:00  
520   2017-03-02 4:00:00  
1104  2017-03-24 4:00:00  
1362  2017-03-15 4:00:00  
1436  2017-03-11 4:00:00  
1562  2017-03-19 4:00:00  
1602  2017-03-17 4:00:00  
2153  2017-03-12 4:00:00  
2297  2017-03-07 4:00:00  
2835  2017-03-28 4:00:00  
2969  2017-03-28 4:00:00  
3332  2017-03-24 4:00:00  
4056  2017-03-28 4:00:00  
4646  2017-03-02 4:00:00  
4868  2017-03-22 4:00:00  
4882  2017-03-25 4:00:00

In [7]:
len(df[df["order_amount"] >= 2000])

63

In [8]:
len(df[df["total_items"] >= 10])

17

In [9]:
df.order_amount.quantile([0.0, .5, .90, .95, .99])

0.00       90.0
0.50      284.0
0.90      531.0
0.95      640.0
0.99    25725.0
Name: order_amount, dtype: float64

_**a. Think about what could be going wrong with our calculation. Think about a better way to evaluate this data.**_

_**Answer to a.**_

As we can see in the output above, the max value in the __order_amount__ column is very high (704000). The standard deviation of the values in this column is also high (41282); therefore, we can claim that there are a couple of outliers in the column, which skew the naively (as mean) calculated average order value. 

We also found out that there are 17 orders in our dataset with 2000 purchased items each. So it can be accurate, but we are talking in this case about a commercial orders. 

On the other hand, all other orders are small and probably were made by individual (private) customers. Hence it would be better to remove the untypically big orders. 

_**b. What metric would you report for this dataset?**_

_**Answer to b.**_

The easiest option is to use the median, which is not that sensitive to outliers as the mean. 

_**c. What is its value?**_


In [10]:
df.order_amount.median()

284.0

_**Answer to c.**_

284 


### Question 2

_**For this question you’ll need to use SQL. Follow this link to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.**_

_**a. How many orders were shipped by Speedy Express in total?**_

_**Answer to a.**_

54 

In [ ]:
SELECT COUNT (OrderID) FROM Orders
WHERE ShipperID = (SELECT ShipperID FROM Shippers
                    WHERE ShipperName = "Speedy Express");

_**b. What is the last name of the employee with the most orders?**_

_**Answer to b.**_

Peacock

In [ ]:
SELECT LastName FROM Employees

WHERE EmployeeID = (SELECT EmployeeID FROM 
						(SELECT COUNT (OrderID), EmployeeID	 
							FROM Orders
                            GROUP BY EmployeeID
                            ORDER BY COUNT (OrderID) DESC
                            LIMIT 1));

_**c. What product was ordered the most by customers in Germany?**_

_**Answer to c.**_

Boston Crab Meat

In [ ]:
SELECT ProductName, SUM(OrderDetails.Quantity) AS Total_orders
FROM Products

JOIN OrderDetails ON OrderDetails.ProductID = Products.ProductID
JOIN Orders ON Orders.OrderID = OrderDetails.OrderID
JOIN Customers ON Customers.CustomerID = Orders.CustomerID

WHERE Customers.Country = "Germany"
GROUP BY Products.ProductName
ORDER BY Total_orders DESC
LIMIT 1;